### PR TITLE
build: fix mypy (local and CI)

### PR DESCRIPTION
**Pull request by chemelli74**
Needs a new release of [midea_local](https://github.com/rokam/midea-local.git) (> 1.1.0) before merge

### CODE REVIEW

**Review by Necroneco — CHANGES_REQUESTED** [review]

File: custom_components/midea_ac_lan/climate.py [hunk 1]
```diff
@@ -173,13 +174,13 @@ def turn_off(self, **kwargs: Any) -> None:
     def set_temperature(self, **kwargs: Any) -> None:
         if ATTR_TEMPERATURE not in kwargs:
             return
-        temperature = float(int((float(kwargs[ATTR_TEMPERATURE]) * 2) + 0.5)) / 2
+        temperature = int(((float(kwargs[ATTR_TEMPERATURE]) * 2) + 0.5) / 2)
```
[the remark this answers is not in the document]
Comment (rokam) [reply custom_components/midea_ac_lan/climate.py:177]: I do agree on this one. The current code just round to .0 or .5, the actual code round to int.

[the remark this answers is not in the document]
Comment (chemelli74) [reply custom_components/midea_ac_lan/climate.py:177]: It was due to temp been and int.
We fixed library so we can revert this change.

File: custom_components/midea_ac_lan/climate.py [hunk 2]
```diff
@@ -488,13 +496,13 @@ def current_temperature(self) -> float | None:
...
```
[the remark this answers is not in the document]
Comment (chemelli74) [reply custom_components/midea_ac_lan/climate.py:500]: Now that lib is fixed we can revert this change

File: custom_components/midea_ac_lan/config_flow.py [hunk 3]
```diff
@@ -194,7 +194,7 @@ async def async_step_login(
                     account=user_input[CONF_ACCOUNT],
                     password=user_input[CONF_PASSWORD],
                 )
-            if await self.cloud.login():
+            elif await self.cloud.login():
```
[the remark this answers is not in the document]
Comment (chemelli74) [reply custom_components/midea_ac_lan/config_flow.py:197]: That could not be possible. if `self.cloud` is None then `self.cloud.login()` cannot work ;-)

File: custom_components/midea_ac_lan/config_flow.py [hunk 4]
```diff
@@ -323,7 +325,7 @@ async def async_step_auto(
                     self.account[CONF_ACCOUNT],
                     self.account[CONF_PASSWORD],
                 )
-            if not await self.cloud.login():
+            elif not await self.cloud.login():
```
[the remark this answers is not in the document]
Comment (chemelli74) [reply custom_components/midea_ac_lan/config_flow.py:328]: Same reply as above

File: custom_components/midea_ac_lan/climate.py [hunk 5]
```diff
@@ -173,13 +174,13 @@ def turn_off(self, **kwargs: Any) -> None:
...
         hvac_mode = kwargs.get(ATTR_HVAC_MODE)
         if hvac_mode == HVACMode.OFF:
             self.turn_off()
         else:
             try:
-                mode = self.hvac_modes.index(hvac_mode.lower()) if hvac_mode else None
+                mode: int = self.hvac_modes.index(hvac_mode.lower()) if hvac_mode else 0
```
[the remark this answers is not in the document]
Comment (chemelli74) [reply custom_components/midea_ac_lan/climate.py:183]: Restore after fixing library

File: custom_components/midea_ac_lan/climate.py [hunk 6]
```diff
@@ -488,13 +496,13 @@ def current_temperature(self) -> float | None:
...
             return
-        temperature = float(int((float(kwargs[ATTR_TEMPERATURE]) * 2) + 0.5)) / 2
+        temperature = int(((float(kwargs[ATTR_TEMPERATURE]) * 2) + 0.5) / 2)
         hvac_mode = kwargs.get(ATTR_HVAC_MODE)
         if hvac_mode == HVACMode.OFF:
             self.turn_off()
         else:
             try:
-                mode = self.hvac_modes.index(hvac_mode.lower()) if hvac_mode else None
+                mode = self.hvac_modes.index(hvac_mode.lower()) if hvac_mode else 0
```
[the remark this answers is not in the document]
Comment (chemelli74) [reply custom_components/midea_ac_lan/climate.py:506]: Restore after fixing library

File: custom_components/midea_ac_lan/climate.py [hunk 7]
```diff
@@ -552,7 +560,7 @@ def current_temperature(self) -> float:
     def set_temperature(self, **kwargs: Any) -> None:
         if ATTR_TEMPERATURE not in kwargs:
             return
-        temperature = float(int((float(kwargs[ATTR_TEMPERATURE]) * 2) + 0.5)) / 2
+        temperature = int(((float(kwargs[ATTR_TEMPERATURE]) * 2) + 0.5) / 2)
```
[the remark this answers is not in the document]
Comment (chemelli74) [reply custom_components/midea_ac_lan/climate.py]: Restore after fixing library

**Comment by chemelli74** [comment]
Create https://github.com/rokam/midea-local/pull/179 to add `set_target_temperature` for `MideaFBDevice` so to use float

**Comment by chemelli74** [comment]
Needs `midea_local` v1.1.2

File: custom_components/midea_ac_lan/config_flow.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Necroneco) [reply custom_components/midea_ac_lan/config_flow.py:197]: No, if self.cloud is None, then self.cloud will be assigned. it works.

change to elif break it

File: custom_components/midea_ac_lan/config_flow.py [hunk 8]
```diff
@@ -333,7 +335,9 @@ async def async_step_auto(
                 CONF_PORT: device.get(CONF_PORT),
                 CONF_MODEL: device.get(CONF_MODEL),
             }
-            if device_info := await self.cloud.get_device_info(device_id):
+            if self.cloud and (
```
Comment (Necroneco) [inline custom_components/midea_ac_lan/config_flow.py:338]: Because of the `if self.cloud is None:` above, the `if self.cloud and` here is not necessary

File: custom_components/midea_ac_lan/config_flow.py [hunk 9]
```diff
@@ -353,9 +357,11 @@ async def async_step_auto(
                             format((PRESET_ACCOUNT[0] ^ PRESET_ACCOUNT[2]), "X"),
                         ).decode("ASCII"),
                     )
-                    if not await self.cloud.login():
+                    if self.cloud and not await self.cloud.login():
```
Comment (Necroneco) [inline custom_components/midea_ac_lan/config_flow.py:360]: same not necessary

File: custom_components/midea_ac_lan/config_flow.py [hunk 10]
```diff
@@ -353,9 +357,11 @@ async def async_step_auto(
...
                         return await self.async_step_auto(error="preset_account")
-                keys = await self.cloud.get_keys(user_input[CONF_DEVICE])
+                keys: dict[int, dict[str, Any]] = {}
+                if self.cloud:
```
Comment (Necroneco) [inline custom_components/midea_ac_lan/config_flow.py:363]: same not necessary

File: custom_components/midea_ac_lan/config_flow.py (hunk 8 above)
Comment (Necroneco) [reply custom_components/midea_ac_lan/config_flow.py:338]: Sorry, my bad. `get_midea_cloud` returns `MideaCloud | None`.
Maybe we should consider raise an error when `get_midea_cloud` still got `None`.

File: custom_components/midea_ac_lan/config_flow.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rokam) [reply custom_components/midea_ac_lan/config_flow.py:197]:
```suggestion
            if self.cloud is not None and await self.cloud.login():
```

File: custom_components/midea_ac_lan/config_flow.py (hunk 8 above)
Comment (rokam) [reply custom_components/midea_ac_lan/config_flow.py:338]: The only way to get_midea_cloud returns None is when the server is invalid. This won´t happen due to the user selection of a valid value. Maybe the solution is to return the user to login step.

**Comment by Necroneco** [comment]
Hello, I made some commits for above conversations. Feel free to modify or drop them if you have different opinions

File: custom_components/midea_ac_lan/config_flow.py (hunk 8 above)
Comment (Necroneco) [reply custom_components/midea_ac_lan/config_flow.py]: Yes, actually `get_midea_cloud` won't return `None`, but the type includes `None`. So simply raise an error to pass mypy check is ok.

Comment (chemelli74) [reply custom_components/midea_ac_lan/config_flow.py]: Is not correct to raise an error just to pass mypy.
Better to fix the logic
If None should never return because the list comes from the library itself, let's fix the return type ;-)

File: custom_components/midea_ac_lan/config_flow.py (hunk 9 above)
Comment (chemelli74) [reply custom_components/midea_ac_lan/config_flow.py]: Did you copy this PR locally, change the code and tru mypy ?

File: custom_components/midea_ac_lan/config_flow.py (hunk 10 above)
Comment (chemelli74) [reply custom_components/midea_ac_lan/config_flow.py]: Same reply as above

**Comment by chemelli74** [comment]
I would rephrase the error and fix the return of the function as we saw it should never be none

**Review by wuwentao — APPROVED** [review]

**Review by wuwentao — APPROVED** [review]
